wasmtime/_memory.py

```diff
@@ -62,6 +62,73 @@ def data_ptr(self, store: Storelike) -> "ctypes._Pointer[c_ubyte]":
         """
         return ffi.wasmtime_memory_data(store._context, byref(self._memory))
 
+    def set_store(self, store: Storelike):
+        """
+        must be called to set the store for highlevel access of memory
+        """
+        self._store=store
+
+    def __getitem__(self, key: int):
+        """
+        provide memory[offset] or memory[start:stop]
+        """
+        if self._store is None:
+            raise RuntimeError("you must call `set_store()` before using highlevel access")
+        data_ptr = self.data_ptr(self._store)
+        size = self.data_len(self._store)
+        if isinstance(key, int):
+            if key>=size:
+                raise IndexError("out of memory size")
+            return data_ptr[key]
+        if not isinstance(key, slice):
+            raise TypeError("key can either be integer index or slice")
+        
+        start, stop, step = key.indices(size)
+        if step!=1:
+            raise ValueError("slice with step is not supported")
+
+        val_size = stop - start
+        value = bytearray(val_size)
+
+        if stop>size:
+            raise IndexError("out of memory size")
+        dst_ptr = (ctypes.c_ubyte * val_size).from_buffer(value)
+        src_ptr = ctypes.addressof((ctypes.c_ubyte*val_size).from_address(ctypes.addressof(data_ptr.contents)+start))
+        ctypes.memmove(dst_ptr, src_ptr, val_size)
+        return value
+
+    def __setitem__(self, key, value: bytearray):
+        """
+        provide setter for memory[key] with slice support
+
+        memory[offset]=value
+        memory[start:stop]=b'hello world'
+        memory[start:stop]=bytearray([1,2,3])
+        """
+        if self._store is None:
+            raise RuntimeError("you must call `set_store()` before using highlevel access")
+        data_ptr = self.data_ptr(self._store)
+        size = self.data_len(self._store)
+        if isinstance(key, int):
+            if key>=size:
+                raise IndexError("out of memory size")
+            data_ptr[key] = value
+        if not isinstance(key, slice):
+            raise TypeError("key can either be integer index or slice")
+
+        start, stop, step = key.indices(size)
+        if step!=1:
+            raise ValueError("slice with step is not supported")
+
+        val_size = len(value)
+        if stop is None:
+            stop=start+val_size
+        if stop-start>val_size or stop>size:
+            raise IndexError("out of memory size")
```

Comment:
I think this logic is the same as above and here, so could this be refactored out?

Reply:
I've made an alternative implementation having which have common parts in  `_get_set_item_common()` 
Do you like this approach or do find it too complex?

https://github.com/muayyad-alsadi/wasm-demos/blob/main/cdb_djp_hash/wasmtime_fast_memory_alt.py

Reply:
I think it would be good to deduplicate this logic, yes, since it's relatively tricky and important to get right.